include/rmm/mr/device/pool_memory_resource.hpp

```diff
@@ -371,20 +396,112 @@ class pool_memory_resource final : public device_memory_resource {
     return std::make_pair(free_size, total_size);
   }
 
+#ifdef CUDA_API_PER_THREAD_DEFAULT_STREAM
+  /**
+   * @brief RAII wrapper for a CUDA event
+   */
+  struct cuda_event {
+    cuda_event(pool_memory_resource<Upstream>* parent) : parent(parent)
+    {
+      auto result = cudaEventCreateWithFlags(&event, cudaEventDisableTiming);
+      assert(cudaSuccess == result);
+    }
+    ~cuda_event() { parent->destroy_event(event); }
+
+    cudaEvent_t event;
+    pool_memory_resource<Upstream>* parent;
+  };
+#endif
+
+  /**
+   * @brief Get the stream associated with `event`.
+   *
+   * @param event The event for which to get the associated stream
+   *
+   * @return The stream associated with `event`
+   */
+  cudaStream_t get_stream(cudaEvent_t event) { return event_streams_[event]; }
+
+  /**
+   * @brief get a unique CUDA event (possibly new) associated with `stream`
+   *
+   * The event is created on the first call, and it is not recorded. If compiled for per-thread
+   * default stream and `stream` is the default stream, the event is created in thread local memory
+   * and is unique per CPU thread.
+   *
+   * @param stream The stream for which to get an event.
+   * @return The CUDA event for `stream`.
+   */
+  cudaEvent_t get_event(cudaStream_t stream)
+  {
+#ifdef CUDA_API_PER_THREAD_DEFAULT_STREAM
+    if (cudaStreamDefault == stream || cudaStreamPerThread == stream) {
```

Comment:
This looks like the kind of thing that will be used in one place and so should be a function. Something like:
```
constexpr bool is_default_stream(cudaStream_t s){ (return s == cudaStreamDefault) or (s == cudaStreamPerThread); }
```

Reply:
It's only used here (which I thought was nice). If I use it again, I'll make it a function.